_blog/2018/soft-multitenancy.md

```diff
@@ -0,0 +1,319 @@
+Pilot should watch for creation of its xDS cache. This is done by starting the Pilot component
+with the additional command line arguments `--appNamespace, ns-1`.  Where *ns-1* is the
+namespace that the tenant’s application will be deployed in. An example snippet from the
+istio-system1.yaml file is included below.
+```yaml
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  name: istio-pilot
+  namespace: istio-system1
+  annotations:
+    sidecar.istio.io/inject: "false"
+spec:
+  replicas: 1
+  template:
+    metadata:
+      labels:
+        istio: pilot
+    spec:
+      serviceAccountName: istio-pilot-service-account
+      containers:
+      - name: discovery
+        image: docker.io/<user ID>/pilot:<tag>
+        imagePullPolicy: IfNotPresent
+        args: ["discovery", "-v", "2", "--admission-service", "istio-pilot", "--appNamespace", "ns-1"]
+        ports:
+        - containerPort: 8080
+        - containerPort: 443
+
+```
+
+### Deploying the tenant application in a namespace
+Now that the cluster administrator has created the tenant's namespace (ex. *istio-system1*) and
+Pilot's service discovery has been configured to watch for a specific application
+namespace (ex. *ns-1*), create the application manifests to deploy in that tenant's specific
+namespace. For example:
+
+```yaml
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: ns-1
+```
+And add the namespace reference to each resource type included in the applications manifest
+file.  For example:
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: details
+  labels:
+    app: details
+  namespace: ns-1
+```
+Although not shown, the application namespaces will also have RBAC settings limiting access
+to certain resources. These RBAC settings could be set by the cluster administrator and/or
+the tenant administrator.
+
+### Using Istioctl commands in a multitenant environment
+When defining [route rules]({{home}}/docs/reference/config/istio.routing.v1alpha1.html#RouteRule)
+or [destination policies]({{home}}/docs/reference/config/istio.routing.v1alpha1.html#DestinationPolicy),
+it is necessary to ensure that the `istioctl` command is scoped to
+the namespace the Istio control plane is running in to ensure the resource is created
+in the proper namespace. Additionally, the rule itself must be scoped to the tenant's namespace
+so that it will be applied properly to that tenant's mesh.  The *-i* option is used to create
+(or get or describe) the rule in the namespace that the Istio control plane is deployed in.
+The *-n* option will scope the rule to the tenant's mesh and should be set to the namespace that
+the tenant's app is deployed in. Note that the *-n* option can be skipped on the command line if
+the .yaml file for the resource scopes it properly instead.
+
+For example, the following command would be required to add a route rule to the *istio-system1*
+namespace:
+```bash
+istioctl –i istio-system1 create -n ns-1 -f route_rule_v2.yaml
+```
```

Comment:
How would this work with istio ingress yaml file, do you also specify -i to specify which ingress controller you are applying the yaml?  Does k8s support multi-tenancy ingress controllers?

Reply:
John,  I was thinking for ingress yaml file, user would use -i to indicate which control plane (that includes istio-ingress) to create the ingress yaml file.    No?  Could you clarify how users would do that?

Reply:
Lin - I am not sure I fully understand your question.   The -i option with istioctl is used to scope the istioctl command to the namespace the associated istio control plane is running in. IOW it creates the resource within the kubernetes namespace associated with the istio control plane.   For route rules and destination policies the control plane is the user of those resources so it should be created in the namespace of the control plane.   Ingress is a bit different.  You don't create an ingress with the istioctl option.   So the "-i" option doesn't apply at all.   It is true that an istio-ingress should be created for each istio control plane in the same namespace as pilot, mixer and CA.   The way we did that in this work was as described in the section above:  "Multiple Istio control planes".  If there is further clarification you would like could you please let us know.

Reply:
Thanks.  I think the clarification would be explaining how ingress will be deployed in the multiple istio control planes environment, like you explained here, e.g. you should create your ingress within the namespace of your Istio control plane and watch out for potential path conflicts yourself?

Reply:
Reworked a bit. Added more content (including mentioning ingress) in "Multiple Istio control planes" section.